toolkit/smart-contracts/offchain/tests/integration_tests.rs

```diff
@@ -432,6 +435,80 @@ async fn governance_action_can_be_initiated_by_non_governance() {
 	run_assemble_and_sign(tx_to_sign, &[GOVERNANCE_AUTHORITY_KEY], &client).await;
 }
 
+#[tokio::test]
+async fn can_upsert_versioned_script_twice() {
+	let image = GenericImage::new(TEST_IMAGE, TEST_IMAGE_TAG);
+	let container = image.start().await.unwrap();
+	let client = initialize(&container).await;
+	let genesis_utxo = run_init_governance(&client).await;
+	let script_id = ScriptId::AlwaysPassingPolicy as u32;
+	let plutus_script_1 = plutus_script![raw_scripts::ALWAYS_FAILING_POLICY, 0u64].unwrap();
+	let plutus_script_2 = plutus_script![raw_scripts::ALWAYS_FAILING_POLICY, 1u64].unwrap();
+	let tx_to_sign = run_upsert_versioned_script(
+		plutus_script_1.clone(),
+		script_id,
+		genesis_utxo,
+		&eve_payment_key(),
+		&client,
+	)
+	.await
+	.unwrap()
+	.unwrap();
+	run_assemble_and_sign(tx_to_sign, &[GOVERNANCE_AUTHORITY_KEY], &client).await;
+
+	let versioned_utxo = find_versioned_utxo(
+		genesis_utxo,
+		ScriptId::AlwaysPassingPolicy,
+		&eve_payment_key(),
+		&client,
+	)
+	.await
+	.unwrap_or_else(|| panic!("Couldn't get versioned utxo"));
+
+	assert_eq!(versioned_utxo.script.unwrap().cbor, plutus_script_1.bytes);
+
+	let tx_to_sign_2 = run_upsert_versioned_script(
+		plutus_script_2.clone(),
+		script_id,
+		genesis_utxo,
+		&eve_payment_key(),
+		&client,
+	)
+	.await
+	.unwrap()
+	.unwrap();
+	run_assemble_and_sign(tx_to_sign_2, &[GOVERNANCE_AUTHORITY_KEY], &client).await;
+
+	let versioned_utxo_2 = find_versioned_utxo(
+		genesis_utxo,
+		ScriptId::AlwaysPassingPolicy,
+		&eve_payment_key(),
+		&client,
+	)
+	.await
+	.unwrap_or_else(|| panic!("Couldn't get versioned utxo"));
+
+	assert_eq!(versioned_utxo_2.script.unwrap().cbor, plutus_script_2.bytes);
+}
+
+#[tokio::test]
+async fn test_upsert_versioned_script_invalid_script_id() {
+	let image = GenericImage::new(TEST_IMAGE, TEST_IMAGE_TAG);
+	let container = image.start().await.unwrap();
+	let client = initialize(&container).await;
+	let genesis_utxo = run_init_governance(&client).await;
+	let plutus_script = plutus_script![raw_scripts::ALWAYS_FAILING_POLICY, 0u64].unwrap();
+	let result = run_upsert_versioned_script(
```

Comment:
Could we melt this one into a test above, because spawning new images and init governance takes some time